### PR TITLE
Add AST load/dump support to PhySL interpreter

### PR DESCRIPTION
**Pull request by parsa**
Providing a `dump-ast=<path.ast>` flag to the interpreter creates an AST dump file from PhySL code.

e.g. `physl code.physl 10 --dump-ast=code.ast`

An AST dump file can be used instead of PhySL code by providing a `load-ast=<path.ast>` flag.
    * e.g. `physl --load-ast=code.ast`

Extras:
* Added a templated overload for `unserialize()` for AST expressions.
* Allow PhySL interpreter users to provide no positional arguments
    * e.g. `physl code.physl`
    * e.g. `physl --load-ast=file.ast`

### CODE REVIEW

File: examples/interpreter/physl.cpp [hunk 1]
```diff
@@ -21,36 +22,97 @@
...
     // Read the file
-    code.assign(std::istreambuf_iterator<char>(code_stream),
-        std::istreambuf_iterator<char>());
+    std::ostringstream str_stream;
+    if (!(str_stream << code_stream.rdbuf()))
+    {
+        HPX_THROW_EXCEPTION(hpx::filesystem_error,
+            "read_user_code",
+            "Failed to read code from the specified file: " + path);
+    }
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: What is the rationale of this change?

File: examples/interpreter/physl.cpp [hunk 2]
```diff
@@ -21,36 +22,97 @@
...
+    std::ifstream str_stream(path, std::ios::binary);
+    auto const start_pos = str_stream.tellg();
+    if (std::streamsize(-1) == start_pos)
+    {
+        HPX_THROW_EXCEPTION(hpx::filesystem_error,
+            "load_ast",
+            "Failed to open the specified file: " + path);
+    }
+
+    if (!str_stream.ignore((std::numeric_limits<std::streamsize>::max)()))
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: This will stop loading if `-1` (`EOF`) is found. What is the rationale of using `.ignore()`?

File: examples/interpreter/physl.cpp [hunk 3]
```diff
@@ -173,43 +234,78 @@ int main(int argc, char* argv[])
...
+    std::vector<phylanx::ast::expression> ast;
+
+    if (vm.count("load-ast") != 0)
     {
-        // Execute code as given directly on the command line
-        user_code = vm["code"].as<std::string>();
-        code_source_name = "<command_line>";
+        auto path = vm["load-ast"].as<std::string>();
+        ast = load_ast(path);
+        code_source_name = "<ast_dump>";
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: we have the file name where we've loaded the AST from. Should we use this here?

File: examples/interpreter/physl.cpp [hunk 4]
```diff
@@ -173,43 +234,78 @@ int main(int argc, char* argv[])
...
+                    read_user_code(vm["transform"].as<std::string>());
+
+                ast = phylanx::ast::transform_ast(
+                    ast, phylanx::ast::generate_transform_rules(transform_rules));
+            }
+
+            // Dump the AST to a file, if requested
+            if (vm.count("dump-ast") != 0)
+            {
+                dump_ast(ast, vm["dump-ast"].as<std::string>());
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: If `--dump-ast` is used without any additional parameters (no file name) we could use the base file name of the input file, just replacing the extension to `.physl-ast` or just `.ast` (or similar).

File: examples/interpreter/physl.cpp [hunk 5]
```diff
@@ -173,43 +219,112 @@ int main(int argc, char* argv[])
...
+                        std::cerr << "physl: command line handling: exception "
+                                     "caught: the required path argument for "
+                                     "option '--dump-ast' is missing\n";
+                        return -1;
+                    }
+                    dump_file = fs::path(positional_args[0])
+                                    .replace_extension("ast")
+                                    .string();
+                }
+                dump_ast(ast, dump_file);
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: You might want to move the ast (as the `dump_ast` function takes the vector by value).

Comment (hkaiser) [reply examples/interpreter/physl.cpp]: Alternatively, change `dump_ast` to take the vector by `const&`.

File: examples/interpreter/physl.cpp [hunk 6]
```diff
@@ -173,43 +219,112 @@ int main(int argc, char* argv[])
...
+            }
+        }
+        catch (std::exception const& e)
+        {
+            std::cout << "physl: exception caught:\n" << e.what() << "\n";
+            return -1;
+        }
```
Comment (hkaiser) [inline examples/interpreter/physl.cpp]: Also, I think the whole code block responsible for creating the AST should go into a separate function.

Comment (hkaiser) [inline examples/interpreter/physl.cpp]: The interpreter can apply an AST transformation. Would it be possible to move the writing of the generated AST further down to enable writing the transformed AST instead of the originally generated one?

Comment (parsa) [reply examples/interpreter/physl.cpp]: Its value is needed. It can't be move. Made it `const&`

File: examples/interpreter/physl.cpp (hunk 5 above)
Comment (parsa) [reply examples/interpreter/physl.cpp]: Its value is needed. It can't be move. Made it `const&`
